addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/handler/MiHomeGatewayHandler.java

```diff
@@ -0,0 +1,575 @@
+                Configuration config = registeredThing.getConfiguration();
+                String code = (String) config.get(MiHomeBindingConstants.CONFIG_GATEWAY_CODE);
+                if (gatewayCode.equals(code) && (!thing.getUID().equals(registeredThing.getUID()))) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public void registerNewGateway() {
+        JsonObject registrationResponse = apiManager.registerGateway(getThing().getLabel(), gatewayCode);
+        if (registrationResponse != null) {
+            JsonObject responseData = registrationResponse.get(JSONResponseConstants.DATA_KEY).getAsJsonObject();
+            gatewayId = responseData.get(DeviceConstants.DEVICE_ID_KEY).getAsInt();
+            setProperties(responseData);
+            updateStatus(ThingStatus.ONLINE);
+        } else {
+            logger.warn(NOT_SUCCESSFULL_REGISTRATION);
+        }
+    }
+
+    private void scheduleUpdateTask(long updateInterval) {
+        if (updateTaskResult != null) {
+            this.updateTaskResult.cancel(true);
+            this.updateTaskResult = null;
+        }
+        /** The refresh thread which is used for checking the "last_seen_at" property of the gateway */
+        Runnable refreshRunnable = new Runnable() {
+            @Override
+            public void run() {
+                logger.debug("Starting the refresh thread...");
+                checkGateway();
+            }
+        };
+        try {
+            scheduler.scheduleAtFixedRate(refreshRunnable, INITIAL_REFRESH_DELAY, updateInterval, TimeUnit.SECONDS);
+        } catch (Exception e) {
+            logger.error("An error occured while trying to check the gateway : {}", e);
+        }
+    }
+
+    /**
+     * Checks when the gateway was last active and updates the thing's channel.
+     * This method is called periodically by a refresh thread.
+     */
+    public void checkGateway() {
+        long timeInterval = 0;
+        /*
+         * Mi|Home REST API does not provide a method for showing a gateway's information.
+         * So in order to check when was the last time it was seen you need to list all
+         * gateways and find the one you want by its id.
+         */
+        JsonObject gateway = getGatewayById(gatewayId);
+        if (gateway != null) {
+            JsonElement lastSeenObj = gateway.get(DeviceConstants.GATEWAY_LAST_SEEN_KEY);
+            gatewayLabel = gateway.get(DeviceConstants.DEVICE_LABEL_KEY).getAsString();
+            thing.setLabel(gatewayLabel);
+            if (!lastSeenObj.isJsonNull()) {
+                String lastSeenString = lastSeenObj.getAsString();
+                String currentDateTimeString = new SimpleDateFormat(MiHomeBindingConstants.LAST_SEEN_PROPERTY_PATTERN)
+                        .format(new Date());
+                timeInterval = getTimeInterval(currentDateTimeString, lastSeenString);
+                if (timeInterval != -1) {
```

Comment:
You could consider user either a Long object or an optional to better indicate presence using -1 as an magic value feels a little bit like the 90-'s